src/plugins/discover/public/application/view_components/utils/use_index_pattern.ts

```diff
@@ -41,6 +45,20 @@ export const useIndexPattern = (services: DiscoverViewServices) => {
     data.indexPatterns,
   ]);
 
+  const fetchDatasetFields = useCallback(
```

Comment:
I think we should handle this in the index pattern service itself. HEre's what im thinking:
1. You create an index pattern with or without schema information depending on dataset type. in the config you can specify async for the type to determine which of the two it is
2. Create a temporary index pattern when the dataset is selected in the configurator. but mark the type as async and defer loading of the schema. This is when you cache the indexpattern with an empty schema
3. while caching, start a promise right after caching it in the index pattern service that when it resolves adds he mappping to the index patttern and set the state to false. See my suggestion in the dataset serivce for reference

Reply:
+1. i think it puts the heavy lifting on the discover plugin. which when we were doing this before the dataset selector there were inaccuracies in the results when we let the plugins ij\nterchange information and try to populate the fields. imo it's a little bit safer for data accuracy if the temp index pattern was created with the fields as a lot of the logic within this area of the code expected the fields to be present if they exist, and non mutating. 

since we update the current index pattern every time the dataset changes this is likely causing an issue where it set index the index pattern and if this is called any time the index pattern is set then its get fields again and sets the index pattern. which then becomes a race condition to which is the right query if the user in the meantime tries to update the language.

usually would be ok merging and fast follows. but the statefulness, multiple requests, and re-renders actually tend to be the source of a bugs that tend to blow up as of late.

Reply:
+1 Updated the code to be async field fetch and not touch use_index_pattern hook.